src/v/kafka/server/group_router.cc

```diff
@@ -83,49 +84,83 @@ auto group_router::route_stages(Request r, FwdFunc func) {
-                      (void)ss::smp::submit_to(
-                        source_shard, [d = std::move(d)]() mutable {
-                            d->set_value();
-                            d.reset();
-                        });
-                  })
-                  .then([f = std::move(stages.result)]() mutable {
-                      return std::move(f);
+                auto dstage = stages.dispatched.then_wrapped(
+                  [rfut = std::move(stages.result)](ss::future<> f) mutable {
+                      auto e = f.failed() ? f.get_exception() : nullptr;
+                      return std::make_tuple(
+                        e, std::make_unique<decltype(rfut)>(std::move(rfut)));
                   });
+
+                return dstage;
             });
+
+          // Immediately above we did not return the stages.result future
+          // directly, rather we wrapped it with a unique_ptr. This means that
+          // the future doesn't receive the cross-shard treatment and is only
+          // safe to use on the coordinator shard, not this one. So next we pass
+          // this future back to the coordinator to unwrap it and return it
+          // unwrapped, so it can be used by this shard (and, as a side effect
+          // the dispatched future is also set).
+          //
+          // Why do we need this second trip rather than just returning the
+          // unwrapped future above?
+          //
+          // Well, this second stage may take a very long time resolve and may
+          // depend on coordination with clients: e.g., the JoinGroup request
+          // won't be responded to until all other current members have also
+          // sent their JoinGroup requests: all the joiners have their responses
+          // held until that moment. Above, the invoke_on is under the influence
+          // of the kafka smp_service_group (_ssg), with a limit of 5,000
+          // outstanding requests on any code to the coordinator: this means
+          // that a group with more than 5,000 members cannot successfully be
+          // joined: at most join 5,000 requests can be outstanding at once, any
+          // more will block on the ssg, and the existing ones won't complete
+          // because they are waiting for the blocked ones: a type of "soft"
+          // deadlock (soft because the server eventually) times out the join
+          // process an the process repeats.
+          //
+          // This is solved by not returning the state.result future directly
+          // above, but wrapping it: this means the returned future resolves as
+          // soon as the dispatch phase completes, freeing up the service group
+          // slot for another request to dispatch. We extract the stage.result
+          // below in a new invoke_on call *without any ssg*: this uses the
+          // default ssg which does not have any limit, avoiding the deadlock.
+
+          return wrapped.then([this,
```

Comment:
am i understand this correctly is `wrapped` variable used to be able to put the comment section in between ?

Reply:
@mmaslankaprv - this was refactored a few times from a more complex change with some additional debugging in between, so it is probably just an artifact of that. I will just chain it directly to the earlier future: the comment can fit in there regardless :).

Reply:
@mmaslankaprv - I've removed the `wrapped` variable, the `.then` just chains on directly from the previous continuation.